output.go

```diff
@@ -0,0 +1,212 @@
+	id        int
+	threshold *stats.Threshold
+}
+
+const expectedDatabaseSchema = `CREATE TABLE IF NOT EXISTS samples (
+		ts timestamptz NOT NULL DEFAULT current_timestamp,
+		metric varchar(128) NOT NULL,
+		tags jsonb,
+		value real
+	);
+	CREATE TABLE IF NOT EXISTS thresholds (
+		id serial,
+		ts timestamptz NOT NULL DEFAULT current_timestamp,
+		metric varchar(128) NOT NULL,
+		tags jsonb,
+		threshold varchar(128) NOT NULL,
+		abort_on_fail boolean DEFAULT FALSE,
+		delay_abort_eval varchar(128),
+		last_failed boolean DEFAULT FALSE
+	);
+	SELECT create_hypertable('samples', 'ts');
+	CREATE INDEX IF NOT EXISTS idx_samples_ts ON samples (ts DESC);
+	CREATE INDEX IF NOT EXISTS idx_thresholds_ts ON thresholds (ts DESC);`
+
+func (o *Output) Start() error {
+	conn, err := o.Pool.Acquire(context.Background())
+	if err != nil {
+		o.logger.WithError(err).Error("TimescaleDB: Couldn't acquire connection")
+	}
+	_, err = conn.Exec(context.Background(), "CREATE DATABASE myk6timescaleDB")
+	if err != nil {
+		o.logger.WithError(err).Debug("TimescaleDB: Couldn't create database; most likely harmless")
+	}
+	_, err = conn.Exec(context.Background(), expectedDatabaseSchema)
+	if err != nil {
+		o.logger.WithError(err).Debug("TimescaleDB: Couldn't create database schema; most likely harmless")
+	}
+
+	for name, t := range o.thresholds {
+		for _, threshold := range t {
+			metric, _, tags := parseThresholdName(name)
+			err = conn.QueryRow(context.Background(),
+				"INSERT INTO thresholds (metric, tags, threshold, abort_on_fail, delay_abort_eval, last_failed) VALUES ($1, $2, $3, $4, $5, $6) RETURNING id",
+				metric, tags, threshold.threshold.Source, threshold.threshold.AbortOnFail, threshold.threshold.AbortGracePeriod.String(),
+				threshold.threshold.LastFailed).Scan(&threshold.id)
+			if err != nil {
+				o.logger.WithError(err).Debug("TimescaleDB: Failed to insert threshold")
+			}
+		}
+	}
+
+	pf, err := output.NewPeriodicFlusher(o.Config.PushInterval, o.commit)
+	if err != nil {
+		return err
+	}
+
+	o.logger.Debug("TimescaleDB: Running!")
+	o.periodicFlusher = pf
+	conn.Release()
+
+	return nil
+}
+
+func (o *Output) commit() {
+	sampleContainers := o.GetBufferedSamples()
+	for _, sc := range sampleContainers {
+		samples := sc.GetSamples()
+		logrus.Debug("TimescaleDB: Committing...")
+		o.logger.WithField("samples", len(samples)).Debug("TimescaleDB: Writing...")
+
+		start := time.Now()
+		batch := &pgx.Batch{}
+		for _, s := range samples {
+			tags := s.Tags.CloneTags()
+			batch.Queue("INSERT INTO samples (ts, metric, value, tags) VALUES ($1, $2, $3, $4)",
+				s.Time, s.Metric.Name, s.Value, tags)
+		}
+
+		for _, t := range o.thresholds {
+			for _, threshold := range t {
+				batch.Queue("UPDATE thresholds SET last_failed = $1 WHERE id = $2",
+					[]interface{}{threshold.threshold.LastFailed, threshold.id},
+					[]pgtype.OID{pgtype.BoolOID, pgtype.Int4OID},
+					nil)
+			}
+		}
+
+		conn, err := o.Pool.Acquire(context.Background())
```

Comment:
I would expect this is quite an expensive operation so maybe moving outside the loop and reusing it for all batch pushes ?

Reply:
Done